### DISCUSSION TITLE
Upgrading lulu to 2.5.1 fails

### DISCUSSION THREAD

**Discussion by 0x0dd4b2** [Homebrew/discussions#4873] [Casks]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: 35746e0a6ba6c3c5cfe56d99f79d9ec9f52ee15f
Last commit: 8 days ago
Core tap JSON: 30 Oct 06:27 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-pytz
  python-packaging
```

### Description of issue

Running `brew upgrade` to upgrade `lulu` from `2.4.3` to `2.5.1` fails with:

```sh
brew upgrade
==> Downloading https://formulae.brew.sh/api/formula.jws.json

==> Downloading https://formulae.brew.sh/api/cask.jws.json

==> Casks with 'auto_updates true' or 'version :latest' will not be upgraded; pass `--greedy` to upgrade them.
==> Upgrading 1 outdated package:
lulu 2.4.3 -> 2.5.1
==> Upgrading lulu
==> Downloading https://github.com/objective-see/LuLu/releases/download/v2.5.1/LuLu_2.5.1.dmg
Already downloaded: /Users/user/Library/Caches/Homebrew/downloads/77b590e364c9c8bd72c933a57c94abad993bd9d33248913e8b7b70ad6c810825--LuLu_2.5.1.dmg
==> Running uninstall script /Applications/LuLu.app/Contents/Resources/LuLu Uninstaller.app/Contents/MacOS/LuLu Uninstaller
Password:
sudo: unable to execute /Applications/LuLu.app/Contents/Resources/LuLu Uninstaller.app/Contents/MacOS/LuLu Uninstaller: Bad CPU type in executable
==> Purging files for version 2.5.1 of Cask lulu
Error: lulu: Failure while executing; `/usr/bin/sudo -E -- /Applications/LuLu.app/Contents/Resources/LuLu\ Uninstaller.app/Contents/MacOS/LuLu\ Uninstaller -uninstall` exited with 1. Here's the output:
sudo: unable to execute /Applications/LuLu.app/Contents/Resources/LuLu Uninstaller.app/Contents/MacOS/LuLu Uninstaller: Bad CPU type in executable
```

### Relevant casks

lulu

**Comment by miccal** [comment]
There is an open issue about this: https://github.com/Homebrew/homebrew-cask/issues/158764

**Reply by 0x0dd4b2** [reply]
~~Searched for an issue but didn't find one, thanks for the reference!~~

I think the issue described in the PR you linked to is different from what my logs above describe. In the PR, uninstalling Lulu succeeds, but then something goes wrong:

> The Lulu cask uninstalls and then installs for upgrades. Uninstalling is accomplished with a script, and in this case it succeeds. However, Homebrew then looks for LuLu.app, which is odd as the uninstall script has removed it.

However, for me the uninstall script fails due to a `Bad CPU type in executable` error. Am I overseeing something here?

I checked the PR you merged and re-tried afterwards, but still getting the same error as described above.

Happy to provide more details / input to help!

**Reply by krehel** [reply]
This looks like an upstream bug.  If you run the Uninstaller manually, it prompts to install Rosetta on an M1 machine.

The Uninstaller has an Xcode project file in the upstream repo where a variable is defined as `VALID_ARCHS = x86_64;`
The main application doesn't seem to have this.
So the Uninstaller looks to just be built for Intel only.

To add extra to this, running uninstall from the menubar from the application itself seems to succeed.  I'll look into it a bit more, but I suspect it just calls the shell script the Uninstaller is wrapped around which is platform agnostic.

I **don't** advise doing this though as it'll leave Homebrew in a confused state by removing an application outside of the standard means.

It may be worth translating the uninstall script into Homebrew `uninstall` stanza language and foregoing the actual uninstaller.  It doesn't appear particularly complex.  I can work on this in the next day.

**Reply by 0x0dd4b2** [reply]
There's an open issue in the upstream repository, https://github.com/objective-see/LuLu/issues/475, but it is quiet old, 26.07.2022. So coming up with a homebrew solution might make sense, considering that there won't come any fix from the project itself.

**Reply by krehel** [reply]
Yeah, I found that same issue (forgot to link it, thanks for including).

Same opinion, we should probably just code this in.  Interesting we haven't seen this come up more often given the circumstances.

But yeah, I'll see what I can come up with in the next day to alleviate the issue.

**Comment by krehel** [comment]
Am still working on this (got sidetracked) but think I have a proof of concept that is working well enough.  Should expect to see an update in the next day or two.

**Reply by 0x0dd4b2** [reply]
Is there maybe something I can help you with in case you got sidetracked, @krehel? Happy to check out the work you've done so far and try to continue it or something alike, let me know! ☺️

**Comment by 0x0dd4b2** [comment]
Lulu recently released version `2.6.2`, see [release notes](https://github.com/objective-see/LuLu/releases/tag/v2.6.2). Within its release notes they mention

> 🚮 Uninstaller now built as native Apple Silicon App

so that I assumed the uninstaller issue was fixed.

Indeed, following the subsequent steps, I was able to remove Lulu from my macbook and re-install it via homebrew.

> **Caution, this causes all your Lulu rules to be deleted, see [here](https://github.com/objective-see/LuLu/issues/581) for additional context!**

Steps executed:

1. Click on the Lulu icon in the menu bar and select `Uninstall Lulu`.
2. Run `brew ls --verbose lulu` to obtain a list of all files related to Lulu that are tracked by homebrew.
3. Remove all files given in the list from 2.
4. Reinstall Lulu via homebrew.

If you want to prevent the loss of all your Lulu rules, there's a new feature [released](https://github.com/objective-see/LuLu/releases/tag/v2.6.0) with `2.6.0`

>  Via the new Rules' submenu you can now export rules (as JSON), or import them

that might be of help. However, I haven't tried it myself yet.

**Reply by 0x0dd4b2** [reply]
cc @krehel (jfyi)
